File: bigdime-alerting-impl/src/main/scala/io/bigdime/alert/impl/swift/SwiftLogTask.scala
package io.bigdime.alert.impl.swift

import java.util.concurrent.Callable

import io.bigdime.libs.client.SwiftAlertClient
import io.bigdime.util.RetryUntilSuccessful
import org.javaswift.joss.client.factory.AccountFactory
import org.javaswift.joss.model.Container
import org.joda.time.format.DateTimeFormat

/**
  * Created by neejain on 2/4/17.
  */
case class SwiftLogTask(val accountFactory: AccountFactory, containerName: String, val sourceName: String, val message: Array[Byte]) extends Callable[Any] {
  final private[swift] val fileNameDtf = DateTimeFormat.forPattern("yyyy-MM-dd")
  var container: Container = _

  @throws[Exception]
  def call: Any = {
    val timestamp = System.currentTimeMillis
    val ts = List[Class[_ <: Throwable]](classOf[Throwable])
    RetryUntilSuccessful(ts)(() => {
      container = accountFactory.createAccount.getContainer(containerName)
      val client = new SwiftAlertClient(container)
      client.writeSegment(getSegmentName(timestamp), message)
      client.writeManifest(getObjectName(timestamp), container.getName + "/" + getPrefixWithDash(timestamp))
    })
  }

  private def getPrefixWithoutDash(timestamp: Long) = {
    sourceName + "." + fileNameDtf.print(timestamp)
  }

  private def getPrefixWithDash(timestamp: Long) = getPrefixWithoutDash(timestamp) + "-"

  private def getSegmentName(timestamp: Long) = getPrefixWithDash(timestamp) + timestamp

  private def getObjectName(timestamp: Long) = {
    "alerts." + sourceName + ".log" + fileNameDtf.print(timestamp)
  }
}

File: bigdime-swift-alerting-impl/src/main/java/io/bigdime/alert/impl/swift/SwiftLogTask.java
package io.bigdime.alert.impl.swift;

import java.util.concurrent.Callable;
import java.util.concurrent.TimeUnit;

import org.javaswift.joss.headers.object.ObjectManifest;
import org.javaswift.joss.instructions.UploadInstructions;
import org.javaswift.joss.model.Container;
import org.javaswift.joss.model.StoredObject;
import org.joda.time.format.DateTimeFormat;
import org.joda.time.format.DateTimeFormatter;

/**
 * Sample filename
 * 
 * bigdime-data-adaptor.2016-01-01.log
 * 
 * Sample segmentNames:
 * 
 * <br>
 * bigdime-data-adaptor.2016-01-01-stamp1
 * 
 * bigdime-data-adaptor.2016-01-01-stamp2
 * 
 * bigdime-data-adaptor.2016-01-01-stamp3
 * 
 * prefixWithoutDash = bigdime-edw-adaptor.2016-01-01
 * 
 * prefixWithDash = prefixWithoutDash + "-"
 * 
 * segmentName = segmentPrefix+ timestamp
 * 
 * objectName = objectPrefix + ".log"
 * 
 * @param source
 * @return
 */
public class SwiftLogTask implements Callable<Object> {
	private final Container container;
	private final String sourceName;
	private final byte[] message;
	final DateTimeFormatter fileNameDtf = DateTimeFormat.forPattern("yyyy-MM-dd");

	public SwiftLogTask(Container _container, String _sourceName, byte[] _message) {
		container = _container;
		sourceName = _sourceName;
		message = _message;
	}

	@Override
	public Object call() throws Exception {
		long timestamp = System.currentTimeMillis();
		writeSegment(timestamp);
		writeManifest(timestamp);
		return null;
	}

	/**
	 * 
	 * @param segmentPrefix
	 * @param timestamp
	 */
	private void writeSegment(long timestamp) {
		String segmentName = getSegmentName(timestamp);
		StoredObject object = container.getObject(segmentName);
		// object.setDeleteAfter(TimeUnit.DAYS.toSeconds(14));
		object.uploadObject(message);
	}

	private void writeManifest(long timestamp) {
		String largeObjectName = getObjectName(timestamp);
		StoredObject object = container.getObject(largeObjectName);
		// object.setDeleteAfter(TimeUnit.DAYS.toSeconds(14));
		UploadInstructions uploadInstructions = new UploadInstructions(new byte[] {});
		uploadInstructions
				.setObjectManifest(new ObjectManifest(container.getName() + "/" + getPrefixWithDash(timestamp)));
		object.uploadObject(uploadInstructions);
	}

	private String getPrefixWithoutDash(long timestamp) {
		String prefix = sourceName + "." + fileNameDtf.print(timestamp);
		return prefix;
	}

	private String getPrefixWithDash(long timestamp) {
		return getPrefixWithoutDash(timestamp) + "-";
	}

	private String getSegmentName(long timestamp) {
		return getPrefixWithDash(timestamp) + timestamp;
	}

	private String getObjectName(long timestamp) {
		return "alerts." + sourceName + ".log" + fileNameDtf.print(timestamp);
		// return getPrefixWithoutDash(timestamp) + ".log";
	}

}
